rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,260 @@
+
+### What metadatas vector adds to the events going through a pipeline
+
+Each time an event goes through a pipeline, vector will create a field (a tag for metrics) `pipelines.${PIPELINE_ID} = ${PIPELINE_VERSION}` in order to be able to track the path taken by an event.
+
+To do so, here are 2 options.
+
+#### By dynamically creating a transform each time a pipeline is used
+
+This step of adding a field should be done with a dynamically created transform and injected in the configuration by the [compiler](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L34).
+Each time a component calls a transform coming from a pipeline, the inputs should be swapped and a dynamic transform created.
+
+```toml
+# in the pipeline load-balancer
+[transforms.do-something]
+...
+
+# in the config
+[sink.output]
+input = ["load-balancer.do-something"]
+...
+```
+
+should be replaced by this equivalent
+
+```
+# in the pipeline load-balancer
+[transforms.do-something]
+...
+
+# injected transform
+[transforms.pre-output]
+input = ["load-balancer.do-something"]
+type = "remap"
+source = '''
+if is_object(.pipelines) {
+  .pipelines = merge(.pipelines, {"load-balancer":"version"})
+} else {
+  .pipelines = {"load-balancer":"version"}
+}
+'''
+
+[sink.output]
+input = ["pre-output"]
+...
+...
+```
+
+To avoid name conflicts with that dynamic transform, this generated transform name could be made by doing a hash of `pre-${NAME}(${INPUTS})`.
+
+#### By adding the field each time an event goes through a pipeline
+
+Adding the ability to a transform to be aware of its name and if it's part of a pipeline by updating `FunctionTransform` and `TaskTransform`:
+
+```rust
+pub trait FunctionTransform: ... {
+  // returning the name and the version
+  fn pipeline(&self) -> Option<(String, String)>;
+  // name of the transform
+  fn name(&self) -> String;
+}
+```
+
+And then adding a `pre_transform` function that would add the pipeline fields
+
+```rust
+pub trait FunctionTransform: ... {
+  fn transform(&mut self, output: &mut Vec<Event>, event: Event) {
+    self.do_transform(output, self.pre_transform);
+  }
+
+  fn pre_transform(&mut self, event: Event) -> Event {
+    // update the event in order to add the field
+  }
+
+  fn transform(&mut self, output: &mut Vec<Event>, event: Event);
+}
+```
+
+This solution would required the core of vector and how it handles transforms but would avoid generatic dynamic transforms.
+It would probably also reduce the performance considering that each transform would have to check if it's part of a pipeline.
```

Comment:
If an approach has performance implications, either positive or negative, I want to validate those before we sink time into the work.